### PR TITLE
skopeo-copy: Allow multiple images to be copied

### LINKED ISSUE #105: skopeo-copy task cannot copy multiple images at once
I am trying to use provided Tasks + cluster resolver instead of ClusterTasks and found one discrepancy between skopeo-copy ClusterTask and the newly-provided Task.
We use skopeo-copy task to copy image under more than one destination image URLs, so we utilize images_url workspace and url.txt file on it. In the ClusterTask, parameter source image URL and destination image URL were optional, i.e. they had default value of an empty string, so if they were not filled, the task would use the url.txt file method. But in the new Task they have no default value. This means they have to be filled, at least with an empty string. This, however, means that the exported_or_fail() function:
https://github.com/openshift-pipelines/task-containers/blob/9d12f07a4b4ebc129c79ce4d54fdda0bdd43b069/scripts/skopeo-common.sh#L19-L23
in the common.sh script makes the task fail because it requires source and destination image urls to be non-empty. It checks using the -z test option:
https://github.com/openshift-pipelines/task-containers/blob/9d12f07a4b4ebc129c79ce4d54fdda0bdd43b069/scripts/common.sh#L24
On the other hand, if we fill in a non-empty string, the task will bypass the url.txt file.

I think the fix would be to mark the source and destination parameters as optional by providing default value "" to them and by removing these params from the following call

https://github.com/openshift-pipelines/task-containers/blob/9d12f07a4b4ebc129c79ce4d54fdda0bdd43b069/scripts/skopeo-common.sh#L19-L23

Support for the task was seemingly added by https://github.com/openshift-pipelines/task-containers/pull/83 but due to the mentioned issue it has probably never worked.

### PR DESCRIPTION
**Pull request by MarianMacik**
fixes #105

### CODE REVIEW

**Review by Aneesh-M-Bhat — CHANGES_REQUESTED** [review]

File: scripts/skopeo-results.sh [hunk 1]
```diff
@@ -15,13 +15,31 @@ function skopeo_inspect() {
...
+    source_digest="$(skopeo_inspect "${PARAMS_SOURCE_IMAGE_URL}" "${PARAMS_SRC_TLS_VERIFY}")"
+    phase "Source image digest '${source_digest}'"
 
-phase "Extracting '${PARAMS_DESTINATION_IMAGE_URL}' destination image digest"
-destination_digest="$(skopeo_inspect "${PARAMS_DESTINATION_IMAGE_URL}" "${PARAMS_DEST_TLS_VERIFY}")"
-phase "Destination image digest '${destination_digest}'"
+    phase "Extracting '${PARAMS_DESTINATION_IMAGE_URL}' destination image digest"
+    destination_digest="$(skopeo_inspect "${PARAMS_DESTINATION_IMAGE_URL}" "${PARAMS_DEST_TLS_VERIFY}")"
+    phase "Destination image digest '${destination_digest}'"
+else
```
Comment (savitaashture) [inline scripts/skopeo-results.sh:26]: Before proceeding directly should we make sure whether workspace is provided or not

**Review by savitaashture — COMMENTED** [review]
:wave: @MarianMacik thank you for the PR
few minor comments

File: test/e2e/e2e-skopeo-copy.bats [hunk 2]
```diff
@@ -76,3 +76,55 @@ declare -rx E2E_SC_PARAMS_DESTINATION_IMAGE_URL="${E2E_SC_PARAMS_DESTINATION_IMA
...
+
+# Testing the skopeo-copy task, copying multiple images using url.txt file, verbose = true
+@test "[e2e] skopeo-copy task copying multiple images from source to destination registry with verbose set to true" {
+    # asserting all required configuration is informed
+	[ -n "${E2E_SC_PARAMS_SOURCE_IMAGE_URL}" ]
+    [ -n "${E2E_SC_PARAMS_DESTINATION_IMAGE_URL}" ]
+    [ -n "${E2E_PARAMS_SRC_TLS_VERIFY}" ]
+    [ -n "${E2E_PARAMS_DEST_TLS_VERIFY}" ]
+
+    E2E_SC_CONFIG_MAP_TEMPLATE_PATH="./test/e2e/resources/configmap-images-template.yaml"
```
[the remark this answers is not in the document]
Comment (MarianMacik) [reply test/e2e/e2e-skopeo-copy.bats:82]: Sorry, fixed now!

File: docs/task-skopeo-copy.md [hunk 3]
```diff
@@ -39,22 +39,28 @@ kubectl create secret docker-registry imagestreams \
 
 Then make sure the Secret is linked with the Service-Account running the `TaskRun`/`PipelineRun`.
 
+## Workspace
+
+| Name         | Optional | Description                                                                                                                                                                                |
+|:-------------|:--------:|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `images_url` |  `true`  | For storing image urls in case of more than one image to copy. It must have a url.txt file at the root path containing a source and a destination image separated by a space on each line. |
```
[the remark this answers is not in the document]
Comment (MarianMacik) [reply docs/task-skopeo-copy.md:84]: Provided.

File: scripts/skopeo-results.sh (hunk 1 above)
Comment (MarianMacik) [reply scripts/skopeo-results.sh:26]: While it is not needed (the task run will fail anyway with the message that url.txt file was not found), I have added to distinguish these 2 scenarios - when the workspace was provided but the file is not present and when the workspace wasn't provided at all.

**Comment by MarianMacik** [comment]
@Aneesh-M-Bhat @savitaashture Everything resolved.

**Comment by Aneesh-M-Bhat** [comment]
LGTM now,
ran the release-tests of skopeo-copy & it works fine.
ran the e2e tests & works as intended.
also ran the below command in openshift cluster 4.16 with operator 1.15.1 installed
```
tkn task start skopeo-copy --use-param-defaults --workspace name=images_url,config=configmap-test-sc --showlog
```
the configmap i used
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: configmap-test-sc
data:
  url.txt: |
    docker://nginx:latest docker://image-registry.openshift-image-registry.svc.cluster.local:5000/anebhat/nginx:latest
    docker://alpine:latest docker://image-registry.openshift-image-registry.svc.cluster.local:5000/anebhat/alpine:latest
    docker://ubuntu:latest docker://image-registry.openshift-image-registry.svc.cluster.local:5000/anebhat/ubuntu:latest

```

**Review by Aneesh-M-Bhat — APPROVED** [review]
/approve

**Pull request merged**